src/Lecture4.hs

```diff
@@ -134,8 +140,45 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+-- I would probably use existing functions, such as:
+-- import GHC.Utils.Misc (split, lengthIs)
+-- But that means I need to change the .cabal file, so I'll copy the
+-- implementation of `split` here and create a custom `hasThreeElements`
+-- function.
+
+split :: Char -> String -> [String]
+split c s = case rest of
+                []     -> [chunk]
+                _:rest -> chunk : split c rest
+  where (chunk, rest) = break (==c) s
+
+hasThreeElements :: [a] -> Bool
+hasThreeElements xs = length xs == 3
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow line = if hasThreeElements listOfValues
+                  then case (product', tradeType', cost') of
+                         (Just p, Just t, Just c) -> Just (Row p t c)
+                         (_, _, _)                -> Nothing
+                  else Nothing
+     where listOfValues = split ',' line
+           (rProduct : rTradeType : rCost : _) = listOfValues  -- This generates a warning*
+           product' = parseProduct rProduct
+           tradeType' = readMaybe rTradeType :: Maybe TradeType
+           cost' = parseCost rCost
```

Comment:
Somehow I have always thought that `do` should be used only in `main`. Not sure where I got that assumption from though. :shrug:
It definitely feels more... "imperative".
Great write out on your website!

Reply:
I think I give up on `guard`. I believe it's a function that takes a Bool, but not sure what it should return?

Reply:
Let's look at the definition of `guard` to understand how it works:

```haskell
guard           :: (Alternative f) => Bool -> f ()
guard True      =  pure ()
guard False     =  empty
```

Here it uses two polymorphic functions:

1. `pure`: it comes from `Applicative` (it's the same as `return` from `Monad`)
2. `empty` comes from the `Alternative`

If we specialise these functions to `Maybe`, the implementation will be equivalent to the following:

```haskell
guard           :: Bool -> Maybe ()
guard True      =  Just ()
guard False     =  Nothing
```

So the function returns `Nothing` on `False` and `Just ()` on `True`.

<hr>

When you write the code like this:

```haskell
cost' <- readMaybe rCost
guard $ cost >= 0
smthElse
```

it's equivalent to the following (after desugaring `do`-notation):

```haskell
readMaybe rCost >>= \cost ->
guard (cost >= 0) >>= \_ ->
smthElse
```

So you can see that `guard` will return either `Just ()` (where we ignore `()` because it's not interesting) or `Nothing`.

`>>=` for `Maybe` either calls the given function or returns `Nothing` when the left argument is `Nothing`. So, in some sense it short-circuits and stops the function when the condition returns `False`.

On other words, it guards the following code and makes sure that the condition is true below `guard` 🙂 

I hope it makes it clearer! 🤗

Reply:
I'm embarrassed... I thought `guard` was a function _you_ would write. Not one that's present in a library.
And thank you very much for the explanation. It's very clear now!